File: planillaventasyegresos/planilla_de_ventas_y_egresos/report/ventas_por_sucursal_y_día/ventas_por_sucursal_y_día.py
import frappe  # type: ignore[import]
import re
from datetime import date, timedelta

# métrica -> (sufijo de fieldname, etiqueta a mostrar)
METRICAS = {
	"Ventas": ("_ventas", "Ventas"),
	"Egresos": ("_egresos", "Egresos"),
	"Saldo Final": ("_saldo", "Saldo"),
}


def _locales():
	"""Locales reales desde las opciones del campo `local` (robusto a renombres)."""
	opts = frappe.get_meta("Caja Diaria").get_field("local").options or ""
	return [o.strip() for o in opts.split("\n") if o.strip()]


def _key(nombre):
	"""fieldname seguro a partir del nombre del local."""
	return re.sub(r"[^a-z0-9]+", "_", (nombre or "").lower()).strip("_")


def _resolve_periodo(filters):
	"""Rellena from_date/to_date según el campo periodo (salvo Personalizado)."""
	periodo = filters.get("periodo") or "Este Mes"
	if periodo == "Personalizado":
		return
	hoy = date.today()
	if periodo == "Hoy":
		filters["from_date"] = filters["to_date"] = hoy
	elif periodo == "Ayer":
		ayer = hoy - timedelta(days=1)
		filters["from_date"] = filters["to_date"] = ayer
	elif periodo == "Esta Semana":
		filters["from_date"] = hoy - timedelta(days=hoy.weekday())
		filters["to_date"] = hoy
	elif periodo == "Este Mes":
		filters["from_date"] = hoy.replace(day=1)
		filters["to_date"] = hoy
	elif periodo == "Mes Pasado":
		primer_dia = hoy.replace(day=1)
		ultimo_mes = primer_dia - timedelta(days=1)
		filters["from_date"] = ultimo_mes.replace(day=1)
		filters["to_date"] = ultimo_mes
	elif periodo == "Este Año":
		filters["from_date"] = hoy.replace(month=1, day=1)
		filters["to_date"] = hoy


def execute(filters=None):
	filters = filters or {}
	_resolve_periodo(filters)
	metric = filters.get("metric") or "Saldo Final"
	locales = _locales()
	return get_columns(metric, locales), get_data(filters, locales)


def get_columns(metric, locales):
	suffix, label_tipo = METRICAS.get(metric, METRICAS["Saldo Final"])

	columns = [{"label": "Fecha", "fieldname": "fecha", "fieldtype": "Date", "width": 110}]
	for s in locales:
		columns.append({
			"label": f"{s} - {label_tipo}",
			"fieldname": f"{_key(s)}{suffix}",
			"fieldtype": "Currency",
			"width": 140,
		})
	columns.append({
		"label": f"Total {label_tipo}",
		"fieldname": f"total{suffix}",
		"fieldtype": "Currency",
		"width": 140,
	})
	return columns


def get_data(filters, locales):
	# condiciones de fecha (separadas para la caja y para el detalle de egresos)
	date_caja = ""
	date_eg = ""
	if filters.get("from_date"):
		date_caja += " AND fecha >= %(from_date)s"
		date_eg += " AND c.fecha >= %(from_date)s"
	if filters.get("to_date"):
		date_caja += " AND fecha <= %(to_date)s"
		date_eg += " AND c.fecha <= %(to_date)s"

	# ventas: soporta subfiltro por medio de pago
	medio = filters.get("medio_pago_venta")
	if medio == "Efectivo":
		ventas_col = "COALESCE(efectivo, 0)"
	elif medio == "Mercado Pago":
		ventas_col = "COALESCE(mercado_pago, 0)"
	elif medio == "Tarjeta de Crédito":
		ventas_col = "COALESCE(tarjeta_credito, 0)"
	elif medio == "Tarjeta de Débito":
		ventas_col = "COALESCE(tarjeta_debito, 0)"
	else:
		ventas_col = "COALESCE(total_ventas, 0)"

	cajas = frappe.db.sql(
		f"""
		SELECT fecha, local,
			{ventas_col}              AS ventas,
			COALESCE(saldo_final, 0)  AS saldo
		FROM `tabCaja Diaria`
		WHERE docstatus < 2 {date_caja}
		""",
		filters,
		as_dict=True,
	)

	# egresos: del detalle, con filtros OPCIONALES por concepto y subconcepto.
	# Los subfiltros son campos YA EXISTENTES del detalle; solo se aplica el que venga seteado.
	eg_cond = ""
	if filters.get("concepto"):
		eg_cond += " AND e.concepto = %(concepto)s"
	for fn in ("tipo_gasto", "tipo_arca", "detalle_municipio", "estado_cheque",
			"local_egreso", "persona", "banco", "nro_patente", "concepto_aporte", "nro_partida"):
		if filters.get(fn):
			eg_cond += f" AND e.{fn} = %({fn})s"

	egresos = frappe.db.sql(
		f"""
		SELECT c.fecha AS fecha, c.local AS local, SUM(COALESCE(e.monto, 0)) AS egresos
		FROM `tabDetalle Egresos Caja` e
		JOIN `tabCaja Diaria` c ON c.name = e.parent
		WHERE c.docstatus < 2 {date_eg} {eg_cond}
		GROUP BY c.fecha, c.local
		""",
		filters,
		as_dict=True,
	)

	keys = {s: _key(s) for s in locales}

	def _ensure(fecha_obj):
		f = str(fecha_obj)
		if f not in pivot:
			pivot[f] = {"fecha": fecha_obj}
			for s in locales:
				pivot[f][f"{keys[s]}_ventas"] = 0
				pivot[f][f"{keys[s]}_egresos"] = 0
				pivot[f][f"{keys[s]}_saldo"] = 0
		return pivot[f]

	pivot = {}
	for r in cajas:
		row = _ensure(r.fecha)
		k = _key(r.local)
		if f"{k}_ventas" in row:
			row[f"{k}_ventas"] += r.ventas
			row[f"{k}_saldo"] += r.saldo
	for r in egresos:
		row = _ensure(r.fecha)
		k = _key(r.local)
		if f"{k}_egresos" in row:
			row[f"{k}_egresos"] += r.egresos

	data = []
	for _, row in sorted(pivot.items()):
		row["total_ventas"] = sum(row.get(f"{keys[s]}_ventas", 0) for s in locales)
		row["total_egresos"] = sum(row.get(f"{keys[s]}_egresos", 0) for s in locales)
		row["total_saldo"] = sum(row.get(f"{keys[s]}_saldo", 0) for s in locales)
		data.append(row)
	return data

File: planillaventasyegresos/planilla_de_ventas_y_egresos/report/ventas_por_sucursal_y_día/ventas_por_sucursal_y_día.js
frappe.query_reports["Ventas por Sucursal y Día"] = {
	filters: [
		{
			fieldname: "periodo",
			label: __("Período"),
			fieldtype: "Select",
			options: ["Este Mes", "Hoy", "Ayer", "Esta Semana", "Mes Pasado", "Este Año", "Personalizado"].join("\n"),
			default: "Este Mes",
			on_change: function(report) {
				let p = report.get_filter_value("periodo");
				let today = frappe.datetime.get_today();
				if (p === "Hoy") {
					report.set_filter_value("from_date", today);
					report.set_filter_value("to_date", today);
				} else if (p === "Ayer") {
					let ayer = frappe.datetime.add_days(today, -1);
					report.set_filter_value("from_date", ayer);
					report.set_filter_value("to_date", ayer);
				} else if (p === "Esta Semana") {
					report.set_filter_value("from_date", frappe.datetime.week_start());
					report.set_filter_value("to_date", frappe.datetime.week_end());
				} else if (p === "Este Mes") {
					report.set_filter_value("from_date", frappe.datetime.month_start());
					report.set_filter_value("to_date", frappe.datetime.month_end());
				} else if (p === "Mes Pasado") {
					let d = frappe.datetime.str_to_obj(today);
					d.setDate(1); d.setMonth(d.getMonth() - 1);
					let ini = frappe.datetime.obj_to_str(d);
					d.setMonth(d.getMonth() + 1); d.setDate(0);
					let fin = frappe.datetime.obj_to_str(d);
					report.set_filter_value("from_date", ini);
					report.set_filter_value("to_date", fin);
				} else if (p === "Este Año") {
					let year = frappe.datetime.str_to_obj(today).getFullYear();
					report.set_filter_value("from_date", year + "-01-01");
					report.set_filter_value("to_date", today);
				}
			},
		},
		{
			fieldname: "from_date",
			label: __("Desde"),
			fieldtype: "Date",
			default: frappe.datetime.month_start(),
			reqd: 1,
			depends_on: "eval:doc.periodo=='Personalizado'",
		},
		{
			fieldname: "to_date",
			label: __("Hasta"),
			fieldtype: "Date",
			default: frappe.datetime.month_end(),
			reqd: 1,
			depends_on: "eval:doc.periodo=='Personalizado'",
		},
		{
			fieldname: "metric",
			label: __("Métrica"),
			fieldtype: "Select",
			options: "Ventas\nEgresos\nSaldo Final",
			default: "Saldo Final",
		},
		{
			fieldname: "medio_pago_venta",
			label: __("Medio de Pago"),
			fieldtype: "Select",
			options: ["", "Efectivo", "Mercado Pago", "Tarjeta de Crédito", "Tarjeta de Débito"].join("\n"),
			depends_on: "eval:doc.metric=='Ventas'",
		},
		{
			fieldname: "concepto",
			label: __("Concepto"),
			fieldtype: "Select",
			options: [
				"",
				"Gastos Generales",
				"Sueldos",
				"Aportes Obra Social",
				"Pagos ARCA",
				"Cuotas de ARCA",
				"Cuotas de ARBA",
				"Cuotas de Municipio",
				"Patentes de Autos",
				"Pagos Alquileres",
				"Pagos Cheques Bco",
				"Gastos Bancarios",
				"Pago Cuotas Préstamos Bco",
				"Retiros",
				"Imp. Provincial",
			].join("\n"),
			depends_on: "eval:doc.metric=='Egresos'",
		},
		{
			fieldname: "tipo_gasto",
			label: __("Tipo de Gasto"),
			fieldtype: "Select",
			options: ["", "Servicios", "Alquiler", "Nafta", "Comida", "Papelería", "Extras"].join("\n"),
			depends_on: "eval:doc.metric=='Egresos' && doc.concepto=='Gastos Generales'",
		},
		{
			fieldname: "tipo_arca",
			label: __("Tipo"),
			fieldtype: "Select",
			options: ["", "IVA", "Ganancias", "IIBB", "Monotributo", "Autónomos", "Otros"].join("\n"),
			depends_on: "eval:doc.metric=='Egresos' && doc.concepto=='Pagos ARCA'",
		},
		{
			fieldname: "detalle_municipio",
			label: __("Detalle"),
			fieldtype: "Select",
			options: ["", "Aportes Seg. e Higiene", "Cuota municipal", "P. Propaganda"].join("\n"),
			depends_on: "eval:doc.metric=='Egresos' && doc.concepto=='Cuotas de Municipio'",
		},
		{
			fieldname: "estado_cheque",
			label: __("Estado"),
			fieldtype: "Select",
			options: ["", "A pagar", "Pagado"].join("\n"),
			depends_on: "eval:doc.metric=='Egresos' && doc.concepto=='Pagos Cheques Bco'",
		},
		{
			fieldname: "local_egreso",
			label: __("Local"),
			fieldtype: "Select",
			options: ["", "CENTRAL", "GRAND BOURG", "JOSE C. PAZ", "SAN MIGUEL", "PILAR", "DEPÓSITO"].join("\n"),
			depends_on: "eval:doc.metric=='Egresos' && doc.concepto=='Pagos Alquileres'",
		},
		{
			fieldname: "nro_patente",
			label: __("Nº Patente"),
			fieldtype: "Data",
			depends_on: "eval:doc.metric=='Egresos' && doc.concepto=='Patentes de Autos'",
		},
		{
			fieldname: "persona",
			label: __("Persona / Nombre"),
			fieldtype: "Data",
			depends_on: "eval:doc.metric=='Egresos' && ['Sueldos','Retiros'].includes(doc.concepto)",
		},
		{
			fieldname: "banco",
			label: __("Banco"),
			fieldtype: "Data",
			depends_on: "eval:doc.metric=='Egresos' && ['Gastos Bancarios','Pago Cuotas Préstamos Bco'].includes(doc.concepto)",
		},
		{
			fieldname: "concepto_aporte",
			label: __("Concepto"),
			fieldtype: "Data",
			depends_on: "eval:doc.metric=='Egresos' && doc.concepto=='Aportes Obra Social'",
		},
		{
			fieldname: "nro_partida",
			label: __("Nº de Partida"),
			fieldtype: "Data",
			depends_on: "eval:doc.metric=='Egresos' && doc.concepto=='Imp. Provincial'",
		},
	],
};
